File: Data Structures/Java/PriorityQueue.java
public class PriorityQueue {
	private static class Node {
		int value;
		Node left, right;
		Node(int val_) {
			value = val_;
			left = null;
			right = null;
		}
	}

	private Node root;
	PriorityQueue() {
		root = null;
	}

	public void push(int data) {
		if(empty())
			root = new Node(data);
		else 
			add(root, data);
	}

	private void add(Node current, int data) {
		if(current.value > data) {
			if(current.left == null)
				current.left = new Node(data);
			else
				add(current.left, data);
		}
		else {
			if(current.right == null)
				current.right = new Node(data);
			else
				add(current.right, data);
		}
	}

	public boolean empty() {
		return root == null;
	}

	public int pop() {
		int result;
		if(empty())
			throw new NullPointerException();
		else if(root.left == null) {
			result = root.value;
			root = root.right;
		}
		else if(root.left.left == null) {
			result = root.left.value;
			root.left = root.left.right;
		}
		else
			result = pop_(root.left);
		return result;
	}

	private int pop_(Node current) {
		int result;
		if(current.left.left == null) {
			result = current.left.value;
			current.left = current.left.right;
		}
		else
			result = pop_(current.left);
		return result;
	}

	public static void main(String[] args) {
		PriorityQueue pq = new PriorityQueue();

		int[] niz = {0, 10, 15, 22, 13, 24, 52, 36, 87, 92, -115, -1, -12, 1};
		for(int el: niz)
			pq.push(el);

		while(!pq.empty())
			System.out.println(pq.pop());
	}
}

File: Data Structures/C++/PriorityQueue.cpp
#include <iostream>

using namespace std;

template<typename T> class Node;
template<typename T> class PriorityQueue;

template<typename T>
class Node {
public:
    T data;
    Node<T>* left;
    Node<T>* right;
    Node(T data)
        : data(data), left(NULL), right(NULL) {}
};

template<typename T>
class PriorityQueue {
private:
    Node<T>* root;
public:
    PriorityQueue()
        : root(NULL) {}

    bool empty() {
        return root == NULL;
    }

    void push(T el) {
        _push(&root, el);
    }

    void show() {
        _show(root);
        cout << endl;
    }

    T pop() {
        if(empty())
            throw "Priority queue is empty.";
        return _pop(&root);
    }
private:
    void _show(Node<T>* curr) {
        if(curr == NULL) {
            cout << ".";
            return;
        }
        cout << "(" << curr->data << " ";
        _show(curr->left);
        cout << " ";
        _show(curr->right);
        cout << ")";
    }

    // For custom classes you have to define operator< !!!
    void _push(Node<T>** curr, T& el) {
        if(*curr == NULL)
            *curr = new Node<T>(el);
        else if(el < (*curr)->data)
            _push(&(*curr)->left, el);
        else
            _push(&(*curr)->right, el);
    }

    T _pop(Node<T>** curr) {
        if((*curr)->left == NULL) {
            Node<T>* abandonedNode = *curr;
            T tmpVal = (*curr)->data;
            *curr = (*curr)->right;
            delete abandonedNode;
            return tmpVal;
        }
        return _pop(&(*curr)->left);
    }
};


int main()
{
    PriorityQueue<int>* pq = new PriorityQueue<int>();
    pq->push(5);
    pq->push(10);
    pq->push(15);
    pq->push(11);
    pq->push(0);
    pq->push(-1);
    pq->push(2);
    pq->push(52);
    pq->push(16);
    pq->push(-1);

    pq->show();
    while(!pq->empty()) {
        cout << pq->pop() << endl;
        pq->show();
    }
    return 0;
}
